crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -207,68 +197,140 @@ impl TryFrom<Type<'_>> for LiteralTypeKind {
     }
 }
 
-impl<'db> IntersectionType<'db> {
-    fn display(&'db self, db: &'db dyn Db) -> DisplayIntersectionType<'db> {
+impl<'a, 'db> IntersectionType<'db> {
+    fn display(&'a self, db: &'db dyn Db) -> DisplayIntersectionType<'a, 'db> {
         DisplayIntersectionType { db, ty: self }
     }
 }
 
-struct DisplayIntersectionType<'db> {
-    ty: &'db IntersectionType<'db>,
+struct DisplayIntersectionType<'a, 'db> {
+    ty: &'a IntersectionType<'db>,
     db: &'db dyn Db,
 }
 
-impl Display for DisplayIntersectionType<'_> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        let mut first = true;
-        for (neg, ty) in self
+impl Display for DisplayIntersectionType<'_, '_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let tys = self
             .ty
             .positive(self.db)
             .iter()
-            .map(|ty| (false, ty))
-            .chain(self.ty.negative(self.db).iter().map(|ty| (true, ty)))
-        {
-            if !first {
-                f.write_str(" & ")?;
-            };
-            first = false;
-            if neg {
-                f.write_str("~")?;
-            };
-            write!(f, "{}", ty.display(self.db))?;
+            .map(|&ty| DisplayMaybeNegatedType {
+                ty,
+                db: self.db,
+                negated: false,
+            })
+            .chain(
+                self.ty
+                    .negative(self.db)
+                    .iter()
+                    .map(|&ty| DisplayMaybeNegatedType {
+                        ty,
+                        db: self.db,
+                        negated: true,
+                    }),
+            );
+        f.join(" & ").entries(tys).finish()
```

Comment:
Nit: I would prefer to keep a for loop here and call `f.entry(&ty)` in the for loop body. It avoids splitting the implementation across multiple types

Reply:
Looks like this doesn't work here; since the `Join` holds a mutable reference to the formatter, you can't `f.write_str(...)` while doing a join, so there's no way to add the `~` for negated types. So we still need `DisplayMaybeNegatedType`, which means it's nicer to just use the iterator and `.entries()`.

I guess the conclusion between this and `DisplayUnionType` is that for loop and `.entry` is good when you're doing conditional stuff; that way you can avoid an extra enum. But it's still a requirement to use `Join` that all the actual writing is done within a per-entry `&dyn Display`; you can't do extra writes to the formatter while joining.